1. Load and analyze the data set

In [1]:
import pandas as pd
from sklearn.datasets import load_breast_cancer

breast_cancer = load_breast_cancer()
print(breast_cancer)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

In [2]:
x = breast_cancer.data
y = breast_cancer.target

print(x.shape, y.shape)

(569, 30) (569,)


In [3]:
data = pd.DataFrame(x, columns = breast_cancer.feature_names)
print(data)

     mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mea

In [4]:
data['class'] = y

data.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension  ...  worst texture  worst perimeter  worst area  \
0                 0.07871  ...          17.33           184.60      2019.0   
1                 0.05667  ...          23.41           158.80      1956.0   
2                 0.05999  ...          25.53           152.50      1709.0   
3                 0.09744  ...          26.50            98.87       567.7   
4                 0.05883  ...          16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  class  
0          0.4601                  0.11890      0  
1          0.2750                  0.08902      0  
2          0.3613                  0.08758      0  
3          0.6638                  0.17300      0  
4          0.2364                  0.07678      0  

[5 rows x 31 columns]

In [5]:
print(data['class'].value_counts())
print(breast_cancer.target_names)

1    357
0    212
Name: class, dtype: int64
['malignant' 'benign']


In [6]:
data.groupby('class').mean()

mean radius  mean texture  mean perimeter   mean area  mean smoothness  \
class                                                                           
0        17.462830     21.604906      115.365377  978.376415         0.102898   
1        12.146524     17.914762       78.075406  462.790196         0.092478   

       mean compactness  mean concavity  mean concave points  mean symmetry  \
class                                                                         
0              0.145188        0.160775             0.087990       0.192909   
1              0.080085        0.046058             0.025717       0.174186   

       mean fractal dimension  ...  worst radius  worst texture  \
class                          ...                                
0                    0.062680  ...     21.134811      29.318208   
1                    0.062867  ...     13.379801      23.515070   

       worst perimeter   worst area  worst smoothness  worst compactness  \
class                                                                      
0           141.370330  1422.286321          0.144845           0.374824   
1            87.005938   558.899440          0.124959           0.182673   

       worst concavity  worst concave points  worst symmetry  \
class                                                          
0             0.450606              0.182237        0.323468   
1             0.166238              0.074444        0.270246   

       worst fractal dimension  
class                           
0                     0.091530  
1                     0.079442  

[2 rows x 30 columns]

2. Split data for training and testing

In [7]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.1, stratify=y, random_state=1)

# stratify ---> for even distribution of data
# random_state ---> specifies split of data, for each random state there's a different set of data


print(x_train.shape, x_test.shape)
print(x.mean(), x_train.mean(), x_test.mean())

(512, 30) (57, 30)
61.890712339519624 61.31637960106119 67.04963097269005


**Using Logistic Regression**

3. Build the model

In [8]:
from sklearn.linear_model import LogisticRegression

model = LogisticRegression()

In [9]:
model.fit(x_train, y_train)

c:\Users\Sachi\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

4. Evaluate the model

In [14]:
from sklearn.metrics import accuracy_score
import numpy as np

In [11]:
# check accuracy on training data
training_set_prediction = model.predict(x_train)
training_set_accuracy_score = accuracy_score(y_train, training_set_prediction)
print(training_set_accuracy_score)

0.951171875


In [12]:
# chaeck accuracy on testing data
testing_set_prediction = model.predict(x_test)
testing_set_accuracy_score = accuracy_score(y_test, testing_set_prediction)
print(testing_set_accuracy_score)

0.9298245614035088


5. Check whether patient has benign or maligant cancer

In [15]:
#for Malignant

test_data = (17.99,10.38,122.8,1001,0.1184,0.2776,0.3001,0.1471,0.2419,0.07871,1.095,0.9053,8.589,153.4,0.006399,0.04904,0.05373,0.01587,0.03003,0.006193,25.38,17.33,184.6,2019,0.1622,0.6656,0.7119,0.2654,0.4601,0.1189)
test_data = np.asarray(test_data)
test_data = test_data.reshape(1, -1)

test_data_prediction = model.predict(test_data)
if test_data_prediction[0] == 0:
    print('Patient has Malignant cancer')
else:
    print('Patient has Benign cancer')


Patient has Malignant cancer


In [16]:
#for Benign

test_data = (13.54,14.36,87.46,566.3,0.09779,0.08129,0.06664,0.04781,0.1885,0.05766,0.2699,0.7886,2.058,23.56,0.008462,0.0146,0.02387,0.01315,0.0198,0.0023,15.11,19.26,99.7,711.2,0.144,0.1773,0.239,0.1288,0.2977,0.07259)
test_data = np.asarray(test_data)
test_data = test_data.reshape(1, -1)

test_data_prediction = model.predict(test_data)
if test_data_prediction[0] == 0:
    print('Patient has Malignant cancer')
else:
    print('Patient has Benign cancer')

Patient has Benign cancer


**Using K-Nearest Neighbours**

3. Build the model

In [17]:
from sklearn.neighbors import KNeighborsClassifier

In [18]:
knn = KNeighborsClassifier(n_neighbors=1)
knn.fit(x_train, y_train)

KNeighborsClassifier(n_neighbors=1)

4. Evaluate the model

In [19]:
#with training data set

knn_training_set_prediction = knn.predict(x_train)
knn_training_set_accuracy_score = accuracy_score(y_train, knn_training_set_prediction)
print(knn_training_set_accuracy_score)

1.0


In [20]:
#with testing data set

knn_testing_set_predicton = knn.predict(x_train)
knn_testing_set_accuracy_score = accuracy_score(y_train, knn_testing_set_predicton)
print(knn_testing_set_accuracy_score)

1.0


5. Check whether ptient has Malignant and Benign cancer

In [21]:
#for Malignant

knn_test_data = (17.99,10.38,122.8,1001,0.1184,0.2776,0.3001,0.1471,0.2419,0.07871,1.095,0.9053,8.589,153.4,0.006399,0.04904,0.05373,0.01587,0.03003,0.006193,25.38,17.33,184.6,2019,0.1622,0.6656,0.7119,0.2654,0.4601,0.1189)
knn_test_data = np.asarray(knn_test_data)
knn_test_data = knn_test_data.reshape(1, -1)

knn_test_data_prediction = knn.predict(knn_test_data)
if knn_test_data_prediction[0] == 0:
    print("Patient has Malignant cancer")
else:
    print("Patient has Benign cancer.")


Patient has Malignant cancer


In [22]:
#for Benign

knn_test_data = (13.54,14.36,87.46,566.3,0.09779,0.08129,0.06664,0.04781,0.1885,0.05766,0.2699,0.7886,2.058,23.56,0.008462,0.0146,0.02387,0.01315,0.0198,0.0023,15.11,19.26,99.7,711.2,0.144,0.1773,0.239,0.1288,0.2977,0.07259)
knn_test_data = np.asarray(knn_test_data)
knn_test_data = knn_test_data.reshape(1, -1)

knn_test_data_prediction = knn.predict(knn_test_data)
if knn_test_data_prediction[0] == 0:
    print("Patient has Malignant cancer")
else:
    print("Patient has Benign cancer.")

Patient has Benign cancer.


**End Result**

Using Logistic regression-
1. Training accuracy score = 0.951171875
2. Testing accuracy score = 0.9298245614035088


Using K-Nearest Neighbours-
1. Training accuracy score = 1.0
2. Testing accuracy score = 1.0